applications/ShapeOptimizationApplication/python_scripts/algorithm_trust_region.py

```diff
@@ -0,0 +1,682 @@
+import data_logger_factory
+import copy
+
+# ==============================================================================
+class AlgorithmTrustRegion(OptimizationAlgorithm):
+    # --------------------------------------------------------------------------
+    def __init__(self, optimization_settings, analyzer, communicator, model_part_controller):
+        default_algorithm_settings = Parameters("""
+        {
+            "name"                          : "trust_region",
+            "max_step_length"               : 1.0,
+            "step_length_tolerance"         : 1e-3,
+            "step_length_reduction_factor"  : 0.5,
+            "max_iterations"                : 10,
+            "far_away_length"               : 2.0,
+            "subopt_max_itr"                : 50,
+            "subopt_tolerance"              : 1e-10,
+            "bisectioning_max_itr"          : 30,
+            "bisectioning_tolerance"        : 1e-2,
+            "obj_share_during_correction"   : 1
+        }""")
+        self.algorithm_settings =  optimization_settings["optimization_algorithm"]
+        self.algorithm_settings.RecursivelyValidateAndAssignDefaults(default_algorithm_settings)
+
+        self.specified_objectives = optimization_settings["objectives"]
+        self.specified_constraints = optimization_settings["constraints"]
+
+        self.analyzer = analyzer
+        self.communicator = communicator
+        self.model_part_controller = model_part_controller
+
+        self.optimization_model_part = model_part_controller.GetOptimizationModelPart()
+        self.design_surface = model_part_controller.GetDesignSurface()
+
+        self.mapper = mapper_factory.CreateMapper(self.design_surface, optimization_settings["design_variables"]["filter"])
+        self.data_logger = data_logger_factory.CreateDataLogger(model_part_controller, communicator, optimization_settings)
+
+        self.geometry_utilities = GeometryUtilities(self.design_surface)
+        self.optimization_utilities = OptimizationUtilities(self.design_surface, optimization_settings)
+
+        self.is_damping_specified = optimization_settings["design_variables"]["damping"]["perform_damping"].GetBool()
+        if self.is_damping_specified:
+            damping_regions = self.model_part_controller.GetDampingRegions()
+            self.damping_utilities = DampingUtilities(self.design_surface, damping_regions, optimization_settings)
+
+    # --------------------------------------------------------------------------
+    def CheckApplicability(self):
+        if self.specified_objectives.size() > 1:
+            raise RuntimeError("Trust-region algorithm only supports one objective function!")
+
+    # --------------------------------------------------------------------------
+    def InitializeOptimizationLoop(self):
+        self.number_of_design_variables = 3*self.design_surface.NumberOfNodes()
+
+        self.x_init = WriteNodeCoordinatesToList(self.design_surface)
```

Comment:
can this be done before `self.model_part_controller.ImportOptimizationModelPart` ?

Reply:
Seems to be unused by the way.

Reply:
Generally not. I corrected it.